diesel/src/r2d2.rs

```diff
@@ -2,6 +2,89 @@
 //!
 //! Note: This module requires enabling the `r2d2` feature
 //!
+//! # Example
+//!
+//! The below snippet is a contrived example emulating a web application,
+//! where one would first initialize the pool in the `main()` function
+//! (at the start of a long-running process). One would then pass this
+//! pool struct around as shared state, which, here, we've emulated using
+//! threads instead of routes.
+//!
+//! ```rust
+//! # include!("doctest_setup.rs");
+//! use diesel::result::Error;
+//! use diesel::pg::PgConnection;
+//! use diesel::prelude::*;
+//! use diesel::r2d2::ConnectionManager;
+//! use diesel::r2d2::Pool;
+//! use std::env;
+//! use std::thread;
+//!
+//! pub type PostgresPool = Pool<ConnectionManager<PgConnection>>;
+//!
+//! pub fn get_connection_pool(pool_size: u32) -> PostgresPool {
+//!     let url = database_url_for_env();
+//!     let manager = ConnectionManager::<PgConnection>::new(url);
+//!     // Refer to the `r2d2` documentation for more methods to use
+//!     // when building a connection pool
+//!     Pool::builder()
+//!         .max_size(pool_size)
+//!         .test_on_check_out(true)
+//!         .build(manager)
+//!         .expect("Could not build connection pool")
+//! }
+//!
+//! pub fn create_user(conn: &mut PgConnection, user_name: &str) -> Result<usize, Error> {
+//!     use schema::users::dsl::*;
+//!
+//!     diesel::insert_into(users)
+//!         .values(name.eq(user_name))
+//!         .execute(conn)
+//! }
+//!
+//! pub fn setup_user_table(conn: &mut PgConnection) {
+//!     diesel::sql_query("DROP TABLE IF EXISTS users CASCADE").execute(conn).unwrap();
+//!     diesel::sql_query("CREATE TABLE users (
+//!         id SERIAL PRIMARY KEY,
+//!         name VARCHAR NOT NULL
+//!     )")
+//!         .execute(conn)
+//!         .unwrap();
+//! }
```

Comment:
Request to reviewers: whether or not I should try and emulate [this logic](https://github.com/diesel-rs/diesel/blob/c6dcd6ba116b86758febf5594c03bfed13aadb3c/diesel/src/doctest_setup.rs#L16) -- i.e. that nothing is ever actually committed.

On one hand, I believe that makes a lot of sense; on the other, since we're using a connection pool (even in a contrived example here where there's only 1 connection), I'm not sure how best to do this -- and furthermore, it's only really a problem if tests run in parallel which I don't think they do but I'm not sure if we want to make that assumption (assuming the Rust ecosystem/some crate one day adds support for running tests in parallel)

Another thing I could do is just run raw SQL queries with `diesel::sql_query` and create and entirely new table `users_table_for_r2d2_module_level_doctest` or something

Reply:
I would probably choose the following strategy:

1. Move the corresponding sql commands in `doctests_setup.rs` in a distinct `setup_schema` function.
2. Use [`r2d2::Builder::connection_customizer`](https://docs.diesel.rs/master/diesel/r2d2/struct.Builder.html#method.connection_customizer) to provide a callback that gets executed once for each connection in the pool + calls the corresponding setup function (that should not be visible to the user, so place a `#` in front of the corresponding lines)
3. Set the connection pool size to 1 (that should not be visible to the user, so place a `#` in front of the corresponding line)

Reply:
Neat! Didn't realize (2) was a thing `r2d2` provided but yeah it's definitely useful :D 

Took a stab at the other two -- open to feedback if it wasn't quite how you envisioned it for sure